CRM: Ensuring users assigned to contacts linked to transactions, invoices, quotes can view those objects

coder-karen opened the pull request:
Fixes https://github.com/Automattic/zero-bs-crm/issues/3443

## Proposed changes:

- This PR makes sure that users who are assigned to contacts are able to see the associated quotes, transactions and invoices linked to those contacts - if 'Assign Ownership' is unchecked from the CRM General Settings page but 'Contact Assignment' is checked.
- Notes: With these assignment settings, a user won't see an invoice, quote, or transaction to which they are not assigned, with the exception of unassigned transactions that only they have created.


### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

https://github.com/Automattic/zero-bs-crm/issues/3443

## Does this pull request change what data or activity we track or use?

No.

## Testing instructions:

You can replicate the issue by:

- Creating a couple of test contacts
- Creating an invoice, transaction, and quote linked to those contacts
- Creating a CRM Contact Manager user
- Making sure 'Contact assignment' is ticked on the General Settings page, and Assign Ownership is unticked: `wp-admin/admin.php?page=zerobscrm-plugin-settings`
- Assigning a contact to that user
- Logging in as that user, and attempting to view the transaction, quote or invoice linked to the assigned contact
- You will see 'You do not have permission to edit this [Transaction, Quote or Invoice]'

To test the fix:

- Apply this PR. Then follow with the above contacts / users and also transaction, quote and invoice, try viewing those again (as the user assigned to the contact). 
- You should be able to view them.
- Try viewing a different transaction, quote or invoice whose assigned contact you are not linked to, and you should see a permissions error.

coder-karen committed:
Ensuring users assigned to contacts linked to transactions, invoices, quotes can see those items

coder-karen committed:
changelog

coder-karen commented on projects/plugins/crm/includes/ZeroBSCRM.Edit.php:
```diff
@@ -170,35 +170,57 @@ public function preChecks(){
+
+				} elseif ( isset( $this->obj['company'][0]['owner'] ) ) {
+					$trans_quote_owner = (int) $this->obj['company'][0]['owner'];
+				// phpcs:disable Generic.WhiteSpace.ScopeIndent.IncorrectExact,Generic.WhiteSpace.ScopeIndent.Incorrect -- this sniff is incorrectly reporting spacing issues.
```
I couldn't find an obvious fix for this sniff issue. This is what I was seeing: 
<img width="792" alt="Screenshot 2024-03-07 at 07 48 19" src="https://github.com/Automattic/jetpack/assets/16754605/dea66f56-31be-48c8-86ce-2847cd6279f5">

gogdzl replied:
Sometimes this old code makes the linter go haywire, I see no problem in disabling the error for this line.

coder-karen committed:
Cast settings checks to int after changing to trict comparison for extra security

coder-karen committed:
Editing two comments to more correctly reflect changes below

coder-karen committed:
Making sure we check if a company is assigned to an invoice before defaulting to that check, prevents being unable to assign the objowner in cases when no assignee

coder-karen committed:
Merge remote-tracking branch 'origin/trunk' into fix/crm-assign-ownership-viewobjects

coder-karen committed:
Changing int casting and moving equality to be loose in two cases - see comments

coder-karen committed:
Merge remote-tracking branch 'origin/trunk' into fix/crm-assign-ownership-viewobjects

gogdzl commented on projects/plugins/crm/includes/ZeroBSCRM.Edit.php:
```diff
@@ -170,35 +170,57 @@ public function preChecks(){
+				}
+
+				// This covers checking owners for assigned contacts or companies in invoices.
+				if ( $this->objTypeID === 4 ) { // phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
```
```suggestion
				if ( $this->objTypeID === ZBS_TYPE_INVOICE ) { // phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
```

gogdzl commented on projects/plugins/crm/includes/ZeroBSCRM.Edit.php:
```diff
@@ -170,35 +170,57 @@ public function preChecks(){
-            $canEditAllContacts = current_user_can('admin_zerobs_customers') && $zbs->settings->get('perusercustomers') == 0;
-            $canGiveOwnership = $zbs->settings->get('usercangiveownership') == 1;
-            $canChangeOwner = ($canGiveOwnership || current_user_can('administrator') || $canEditAllContacts);
+			if ( $obj_owner > 0 && $obj_owner !== $current_user_id ) {
```
```suggestion
			if ( ( $obj_owner > 0 && $obj_owner !== $current_user_id ) || $obj_owner === -1 ) {
```

gogdzl replied:
This change would prevent unassigned contacts to be viewed by all contact managers.
I'm suggesting this because this is our current behavior, and in my humble opinion, we should not change it.

I think that  comparing to `-1` is enough to make the assertion that the contact is unassigned.

coder-karen replied:
This behaviour hasn't been changed - prior to this PR CRM Contact Managers could view all unassigned contacts as well, so that hasn't been changed here. 

That is with 'Contact Assignment' checked and 'Assign Ownership' unchecked. It's the same if both are unchecked, and checked, both before and after this PR, for me, too. It should work as is currently anyway because if `$obj_owner` is -1 (unassigned), then it will return true for `$obj_owner !== $current_user_id`. I do see however that I changed this to be a strict comparison, and that should have been changed back with the other comparison reverts (as there is the chance some results will be strings not ints). I'll change that now - 0f0a197d7bd729641daf45ba380f153df72b32a5. I'm not sure if that makes any difference in your test case though, let me know if so.

gogdzl replied:
I apologize for the confusion in my earlier message. Rectifying, only transactions, quotes, and invoices linked to unassigned contacts  were made accessible to contact managers in this branch.

The interaction with transactions/invoices/quotes linked to assigned contacts was preserved.

This is what happens:

**TRUNK (as a contact manager)**
Unassigned Contact (I can see the contact):
![image](https://github.com/Automattic/jetpack/assets/37049295/bc47348e-a366-4b62-aa5b-ff674b7f7291)

The same contact's transaction:
![image](https://github.com/Automattic/jetpack/assets/37049295/21135a61-b696-4fa2-9bba-06a0271c3ef9)
![image](https://github.com/Automattic/jetpack/assets/37049295/aecc5f0d-c9fd-41ea-8226-ab72423b3778)


**fix/crm-assign-ownership-viewobjects (as that same contact manager)**
Unassigned Contact (I can see the contact):
![image](https://github.com/Automattic/jetpack/assets/37049295/b08423a5-0da8-42e7-b09f-a06ea1f22919)

The same contact's transaction:
![image](https://github.com/Automattic/jetpack/assets/37049295/1253a12e-e6b4-4e36-8851-89f4f69ea7e5)
![image](https://github.com/Automattic/jetpack/assets/37049295/adefa550-1ced-47c1-8a9b-8fb5a3b33ba1)

coder-karen replied:
I do actually see what you mean now, thanks! 

The behaviour in trunk is slightly more nuanced, here's what I found:

Trunk: Very inconsistent behaviour - some older transactions, invoices, quotes, I can't view
 regardless of being unassigned, assigned to a contact with no owner, and newer ones I can view in the same circumstances plus being linked to a contact with owner (regardless of owner). 

Branch: Behaviour is consistent. Except - transactions and invoices created by the contact manager and unassigned or assigned to a contact with no owner are viewable (previously I thought it was just when unassigned), or the same objects created by admin and assigned to a contact with no owner.

The one thing there that looks like it needs fixing is the situation you mentioned - the object linked to a contact who doesn't have an owner, though there isn't a clear description of what this should have been previously. It just makes sense to fix. And your fix of course does work! I'll remove the strict equality check as `$obj_owner` can apparently be a string in some situations. Commit added here: 0e8fe5b59b5dde1d44cdd86f78627ab9f031d28b

gogdzl reviewed (changes requested):
Great job tackling this!
I left some comments, being the change of behavior a blocker for me.

Also, you get bonus points if you chop this: `if ($zbs->isDAL3()){`

tbradsha commented:
> Also, you get bonus points if you chop this: `if ($zbs->isDAL3()){`

Spoken like a true janitor!

coder-karen committed:
Removing strict equality check

coder-karen committed:
Removing DAL3 check

coder-karen committed:
Merge remote-tracking branch 'origin/trunk' into fix/crm-assign-ownership-viewobjects

coder-karen committed:
Update projects/plugins/crm/includes/ZeroBSCRM.Edit.php

Co-authored-by: gogdzl <37049295+gogdzl@users.noreply.github.com>

coder-karen commented:
@gogdzl Thanks for the review! I've removed that DAL check and added your constant addition. Left a comment here regarding viewing unassigned contacts: https://github.com/Automattic/jetpack/pull/36239#discussion_r1519579681

gogdzl commented on projects/plugins/crm/includes/ZeroBSCRM.Edit.php:
```diff
@@ -164,19 +164,16 @@ public function preChecks(){
         global $zbs;
 
-        // only do this stuff v3.0+
-        if ($zbs->isDAL3()){
```
Nice!

gogdzl reviewed (dismissed):
I left a comment here: https://github.com/Automattic/jetpack/pull/36239#discussion_r1520795083

If this behavior is intended, then we're good to go! I've left the PR approved, assuming this is the desired behavior for the CRM.

coder-karen committed:
Adding check for transactions and invoices with assigned contacts but no owner (no obj_owner), preventing those from being visible

coder-karen commented:
@gogdzl followed up in the thread so a re-review makes sense now, thanks!

gogdzl reviewed (approved):
LGTM!
